pep-9999.rst

```diff
@@ -0,0 +1,675 @@
+precise types for ``open(…)``. While this approach works for standard
+library functions, it’s unsustainable in general: it’s not reasonable to
+expect 3rd party library authors to maintain plugins for N different
+type checkers, for example.
+
+We propose adding *Literal types* to address these gaps.
+
+Specification
+=============
+
+Core Semantics
+--------------
+
+This section outlines the baseline behavior of literal types.
+
+Core behavior
+'''''''''''''
+
+Literal types indicate a variable has a specific and
+concrete value. For example, if we define some variable ``foo`` to have
+type ``Literal[3]``, we are declaring that ``foo`` must be exactly equal
+to ``3`` and no other value.
+
+Given some value ``V`` that is a member of type ``T``, the type
+``Literal[V]`` shall be treated as a subtype of ``T``. For example,
+``Literal[3]`` is a subtype of ``int``.
+
+All methods from the parent type will be directly inherited by the
+literal type. So, if we have some variable ``foo`` of type ``Literal[3]``
+it’s safe to do things like ``foo + 5`` since ``foo`` inherits int’s
+``__add__`` method. The resulting type of ``foo + 5`` is ``int``.
+
+This “inheriting” behavior is identical to how we handle NewTypes.
+
+Equivalence of two Literals
+'''''''''''''''''''''''''''
+
+Two types ``Literal[A]`` and ``Literal[B]`` are equivalent when
+both of the following conditions are true:
+
+1. ``type(A) == type(B)``
+2. ``A == B``
+
+For example, ``Literal[20]`` and ``Literal[0x14]`` are equivalent.
+However, ``Literal[0]`` and ``Literal[False]`` is *not* equivalent
+despite that ``0 == False`` evaluates to 'true' at runtime: ``0``
+has type ``int`` and ``False`` has type ``bool``.
+
+Shortening unions of literals
+'''''''''''''''''''''''''''''
+
+Literals are parameterized with one or more value. When a Literal is
+parameterized with more then one value, it's treated as exactly equivalent
+to the union of those types. That is, ``Literal[V1, V2, V3]`` is equivalent
+to ``Union[Literal[V1], Literal[V2], Literal[V3]]``.
+
+This shortcut helps make writing signatures for functions that accept
+many different literals more ergonomic — for example, functions like
+``open(...)``::
+
+   # Note: this is a simplification of the true type signature.
+   _PathType = Union[str, bytes, int]
+
+   @overload
+   def open(path: _PathType, 
+            mode: Literal["r", "w", "a", "x", "r+", "w+", "a+", "x+"],
+            ) -> IO[Text]: ...
+   @overload
+   def open(path: _PathType, 
+            mode: Literal["rb", "wb", "ab", "xb", "r+b", "w+b", "a+b", "x+b"],
+            ) -> IO[bytes]: ...
+
+   # Fallback overload for when the user isn't using literal types
+   @overload
+   def open(path: _PathType, mode: str) -> IO[Any]: ...
+
+**Note:** Literals **must** be parameterized with at least one type.
+Types like ``Literal[]`` or ``Literal`` are illegal.
+
+Legal and illegal parameterizations
+-----------------------------------
```

Comment:
I decided to rename the final section from "Miscellaneous interactions" to "Interactions with other types and features" but decided to otherwise keep the sections the same. I like the flow of starting with an relatively minimal sketch or skeleton of the behavior, then diving into the details moving from the inside-out: discuss first what can go inside a Literal, next when can we infer a Literal, and finally how do Literals (once constructed) interact with other things.

I did remove one level of nesting though -- I don't think shoving 80% of the doc into a "specifications" section was really necessary.